### PR TITLE
Implementation of ble-scan method with returning discovered devices

### LINKED ISSUE #13: Support Parsing BLE Advertisement Data #165
TH should allow scanning and parsing of BLE Advertisement data
CSG Custom controller can be made to either:
Report back parsed Advertisement packets via RPC hooks (Or)
Report back Advertisement logs which the TH can then parse

### PR DESCRIPTION
**Pull request by nirav1911**
Problem
Updated ble_scan method so it can return the discovered devices through Json.

Summary of Changes
Retrieved the list of devices from BleManager as well as checked if the OS is supported for that or not. Currently supporting only Linux OS.

It will also filter out non-CHIP devices.
Handled :https://github.com/chip-csg/chip-certification-tool-backend/issues/50

Fixed #13

### CODE REVIEW

File: src/controller/python/chip-device-ctrl.py [hunk 1]
```diff
@@ -617,20 +624,68 @@ def emptyline(self):
...
+            device_detail['discriminator'] = devIdInfo.discriminator
+            device_detail['vendor_id'] = devIdInfo.vendorId
+            device_detail['product_id'] = devIdInfo.productId
+            if device.ServiceData:
+                for advuuid in device.ServiceData:
+                    device_detail['adv_uuid'] = str(advuuid)
+            device_list.append(device_detail)
+    return device_list
+
+def ble_connect(discriminator: int, pin_code: int, node_id: int) -> string:
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:662]: Return type should be Dict [str, Any] ?

File: src/controller/python/chip-device-ctrl.py [hunk 2]
```diff
@@ -617,20 +624,68 @@ def emptyline(self):
...
+    try:
+        __check_supported_os()
+        device_manager.devCtrl.ConnectBLE(discriminator, pin_code, node_id)
+        return __get_response_dict(status = StatusCodeEnum.SUCCESS)
+    except exceptions.ChipStackException as ex:
+        return __get_response_dict(status = StatusCodeEnum.FAILED, error = str(ex))
+    except Exception as e:
+        return __get_response_dict(status = StatusCodeEnum.FAILED, error = str(e))
+
+def ip_connect(ip_address: string, pin_code: int, node_id: int) -> string:
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:672]: Return type should be Dict [str, Any] ?

File: src/controller/python/chip-device-ctrl.py [hunk 3]
```diff
@@ -639,6 +694,23 @@ def start_rpc_server():
             print("\nKeyboard interrupt received, exiting.")
             sys.exit(0)
 
+def __get_response_dict(status: StatusCodeEnum, result: Optional[Dict[Any, Any]] = None, error:Optional[str] = None) -> Dict [Any, Any]:
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:697]: We can narrow down the return type to Dict [str, Any]?

File: src/controller/python/chip-device-ctrl.py [hunk 4]
```diff
@@ -639,6 +694,23 @@ def start_rpc_server():
...
+    if error is not None:
+        return { "status" : status.value, "error" :f'exception {error}' }
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:699]: Nit: "status", "result" could be str enums

File: src/controller/python/chip-device-ctrl.py [hunk 5]
```diff
@@ -639,6 +694,23 @@ def start_rpc_server():
...
+    else:
+        if result is not None:
+            return { "status" : status.value, "result": result}
+        else:
+            return { "status" : status.value}
+
+def __check_supported_os()-> bool:
+    if platform.system() == 'Darwin':
+        raise Exception(platform.system() + " not supported")
+    elif sys.platform.startswith('linux'):
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:709]: Should we worry about upper-case, lower-case issues here?

File: src/controller/python/chip-device-ctrl.py (hunk 3 above)
Comment (wahajsyed) [reply src/controller/python/chip-device-ctrl.py:697]: Nit: '= None' is not required when marked with Optional AFAIK

File: src/controller/python/chip-device-ctrl.py [hunk 6]
```diff
@@ -617,20 +624,68 @@ def emptyline(self):
...
+            device_detail['id'] = str(device.device_id)
+            device_detail['rssi'] = str(device.RSSI)
+            device_detail['address'] = str(device.Address)
+            device_detail['pairing_state'] = devIdInfo.pairingState
+            device_detail['discriminator'] = devIdInfo.discriminator
+            device_detail['vendor_id'] = devIdInfo.vendorId
+            device_detail['product_id'] = devIdInfo.productId
+            if device.ServiceData:
+                for advuuid in device.ServiceData:
+                    device_detail['adv_uuid'] = str(advuuid)
```
Comment (wahajsyed) [inline src/controller/python/chip-device-ctrl.py:658]: Is device.ServiceData a list here? Should `device_detail['adv_uuid']` be a list as well?

**Review by wahajsyed — COMMENTED** [review]
Added Some comments

File: src/controller/python/chip-device-ctrl.py (hunk 6 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py:658]: I used the logic what they have described in blemanager.

File: src/controller/python/chip-device-ctrl.py (hunk 5 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved.

File: src/controller/python/chip-device-ctrl.py (hunk 4 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved.

File: src/controller/python/chip-device-ctrl.py (hunk 3 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved.

File: src/controller/python/chip-device-ctrl.py (hunk 2 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved

File: src/controller/python/chip-device-ctrl.py (hunk 1 above)
Comment (nirav1911) [reply src/controller/python/chip-device-ctrl.py]: Resolved

**Review by wahajsyed — APPROVED** [review]
LGTM

**Review by mikaelhm — APPROVED** [review]

**Pull request merged**